test/functional/eval/fnamemodify_spec.lua

```diff
@@ -36,4 +36,187 @@ describe('fnamemodify()', function()
+    end)
+  end)
+
+  describe('advanced examples from ":help filename-modifiers"', function()
+    local filename = "src/version.c.gz"
+
+    it(':e', function()
+      eq('gz', fnamemodify(filename, ':e'))
+    end)
+
+    it(':e:e', function()
+      eq('c.gz', fnamemodify(filename, ':e:e'))
+    end)
+
+    it(':e:e:e', function()
+      eq('c.gz', fnamemodify(filename, ':e:e:e'))
+    end)
+
+    it(':e:e:r', function()
+      eq('c', fnamemodify(filename, ':e:e:r'))
+    end)
+
+    it(':r', function()
+      eq('src/version.c', fnamemodify(filename, ':r'))
+    end)
+
+    it(':r:e', function()
+      eq('c', fnamemodify(filename, ':r:e'))
+    end)
+
+    it(':r:r', function()
+      eq('src/version', fnamemodify(filename, ':r:r'))
+    end)
+
+    it(':r:r:r', function()
+      eq('src/version', fnamemodify(filename, ':r:r:r'))
+    end)
+  end)
+
+  describe('modify with :h', function()
+    it('handles bare filenames', function()
+      eq('.', fnamemodify('hello.txt', ':h'))
+    end)
+
+    it('handles paths', function()
+      eq('path/to', fnamemodify('path/to/hello.txt', ':h'))
+    end)
+  end)
+
+  describe('modify with :t', function()
+    it('handles bare filenames', function()
+      eq('hello.txt', fnamemodify('hello.txt', ':t'))
+    end)
+
+    it('handles paths', function()
+      eq('hello.txt', fnamemodify('path/to/hello.txt', ':t'))
+    end)
+  end)
+
+  describe('modify with :r', function()
+    it('handles bare filenames', function()
+      eq('hello', fnamemodify('hello.txt', ':r'))
+    end)
+
+    it('handles paths', function()
+      eq('path/to/hello', fnamemodify('path/to/hello.txt', ':r'))
+    end)
+  end)
+
+  describe('modify with :e', function()
+    it('handles bare filenames', function()
+      eq('txt', fnamemodify('hello.txt', ':e'))
+    end)
+
+    it('handles paths', function()
+      eq('txt', fnamemodify('path/to/hello.txt', ':e'))
+    end)
+  end)
+
+  describe('modify with regex replacements', function()
+    it('handles a simple s///', function()
+      eq('content-there-here.txt', fnamemodify('content-here-here.txt', ':s/here/there/'))
+    end)
+
+    it('handles global', function()
+      eq('content-there-there.txt', fnamemodify('content-here-here.txt', ':gs/here/there/'))
+    end)
+  end)
+
+  it('modify with shell escape', function()
+    eq("'hello there! quote '\\'' newline\n'", fnamemodify("hello there! quote ' newline\n", ':S'))
+  end)
+
+  describe('combining :e and :r', function()
+    describe('single-extension filename', function()
+      it('handling a.c :e', function() eq('c', fnamemodify('a.c', ':e')) end)
+      it('handling a.c :e:e', function() eq('c', fnamemodify('a.c', ':e:e')) end)
+      it('handling a.c :e:e:r', function() eq('c', fnamemodify('a.c', ':e:e:r')) end)
+      it('handling a.c :e:e:r:r', function() eq('c', fnamemodify('a.c', ':e:e:r:r')) end)
+    end)
+
+    describe('multiple-extension filename', function()
```

Comment:
Thanks for the thoroughness!

This could be an `it()` , and all of the nested `it()` cases in here can be in just that 1 test.  Each `it()` starts a new nvim process and there's not really much need for that here, so it's better to avoid that cost in test time.

```lua
it('multiple-extension filename', function()
  -- handling a.spec.rb :e
  eq('rb', fnamemodify('a.spec.rb', ':e:r'))
  -- handling a.spec.rb :e:r
  eq('rb', fnamemodify('a.spec.rb', ':e:r'))
  -- handling a.spec.rb :e:e:r
  eq('spec', fnamemodify('a.spec.rb', ':e:e:r'))
  -- handling a.spec.rb :e:e:r:r
  eq('spec', fnamemodify('a.spec.rb', ':e:e:r:r'))
  -- handling a.b.spec.rb :e:e:r
  eq('spec', fnamemodify('a.b.spec.rb', ':e:e:r'))
  -- handling a.b.spec.rb :e:e:e:r
  eq('b.spec', fnamemodify('a.b.spec.rb', ':e:e:e:r'))
  -- handling a.b.spec.rb :e:e:e:r:r
  eq('b', fnamemodify('a.b.spec.rb', ':e:e:e:r:r'))

  -- handling a.b.spec.rb :r:e
  eq('spec', fnamemodify('a.b.spec.rb', ':r:e'))
  -- handling a.b.spec.rb :r:r:e
  eq('b', fnamemodify('a.b.spec.rb', ':r:r:e'))
end)
```

Same for the other cases in this PR.

Reply:
Sure, shall do :+1:

Out of interest (as I'd originally written these as unit tests with `itp`), is there much difference in terms of execution between the functional test's `it` and the unit test's `itp` ?

Reply:
> Out of interest (as I'd originally written these as unit tests with `itp`), is there much difference in terms of execution between the functional test's `it` and the unit test's `itp` ?

Not really. The unit tests (`itp`) start a new process IIRC.

NB: `itp` is our wrapper around busted's `it`, which is needed only for unit tests (not functional tests).

Reply:
Cool, thanks for the explanation :)

Reply:
Yes, `itp` is needed/used to make C functions available for unit tests (via LuaJIT's ffi).